### PR TITLE
refactor(auth): Various build improvements and minor refactoring

### PR DESCRIPTION
**Pull request by rpatterson**
Revert the commits that changed the components that refer to the logged in user to use
the response from the actions API endpoint.  IOW, those components now, once again, use
the Bearer token instead.  Preserves all the other improvements made to the code along
the way.

The new branch for the `plone.restapi` includes the API changes to the PAS plugin needed
to set both the Plone classic session cookie and the Bearer token cookie that Volto can
optionally use to initialize authentication without using the Volto login component.

All this put together means that:
- the Volto components related to logged in users now behave largely as they did before
  in terms of interpreting state
- logging in to or out of either classic or Volto also does so in the other

One edge case that remains is when logging in via the Volto UI component as a user from
the Zope root `acl_users` (e.g. `admin` or `SITE_OWNER_NAME`), they aren't logged into
Plone classic.  The other direction is fine, logging in as `admin` to Plone classic
results in a new request to the Volto UI being logged in.  Digging into that now.

These changes do not address any of the issues that happen when logged in via the ZMI at
the Zope root.  A user logged in that way will still appear logged in to classic Plone,
but won't be recognized by Volto components as logged in but API responses will still
reflect their `Manager` access rights.  Tackling that is coming up.

This PR supersedes both #2535 #2687.

### CODE REVIEW

**Comment by rpatterson** [comment]
I'm trying to see if I can't reproduce the CI check failures locally.  I'll request review later when I think it's ready.

**Comment by rpatterson** [comment]
K, ready for review now.

File: src/helpers/Utils/Utils.js
Comment (tiberiuichim) [inline src/helpers/Utils/Utils.js:196]: I think this can be a helper inside the actions reducer, I don't see its utility exposed as a Volto global helper.

Comment (rpatterson) [reply src/helpers/Utils/Utils.js:196]: I put it in `Utils` because there are many cases in CMF's and Plone's data shapes where there are arrays of objects with `id` properties.  So the function isn't specific to actions and there's likely to be fruitful usage outside of actions.

File: src/actions/actions/actions.js
Comment (tiberiuichim) [inline src/actions/actions/actions.js:15]: Why is this style of action needed? I don't know if this is a default feature of react-redux or it comes to us thanks to thunks support, but we're already able to treat the actions as async and wait their result, see for ex https://github.com/plone/volto/blob/de3cb700c2ba44e7b19c3ebafc98d2bbba0852e9/src/components/manage/Widgets/SelectAutoComplete.jsx#L150

Comment (rpatterson) [reply src/actions/actions/actions.js:15]: I think @sneridagh is the one to weigh in on this, no?

File: src/helpers/Utils/Utils.js
Comment (tiberiuichim) [reply src/helpers/Utils/Utils.js:196]: If it's a global helper, it needs a better name, otherwise it's hard to remember.

Comment (rpatterson) [reply src/helpers/Utils/Utils.js:196]: Such as?

File: api/buildout.cfg
Comment (tiberiuichim) [inline api/buildout.cfg:21]: I think this changes shouldn't be committed here. You can add them as instructions in the PR comments, but with them here it makes it hard to merge. I understand they're prerequisite for the PR.

**Comment by rpatterson** [comment]
For whoever merges/releases this, see [details on what changes/PRs/releases in other packages are required for this PR](https://github.com/plone/volto/pull/2959/commits/f145c3d50e1b6ad0f13b90a233eb0446b0dab798)

File: api/buildout.cfg
Comment (rpatterson) [reply api/buildout.cfg]: Done.

**Comment by tiberiuichim** [comment]
@rpatterson So, as far as I can see in this PR, the changes are "trivial": the main change is the introduction of the userLoggedIn as a signal that a user is authenticated, instead of having it based on the existence of the token. All the other changes are cosmetic, right?

(It's another bug, but the existence of the token doesn't make the user logged in, the token could be "expired").

**Comment by tiberiuichim** [comment]
Needs https://github.com/plone/plone.restapi/pull/1303 and https://github.com/plone/plone.volto/pull/39

**Comment by rpatterson** [comment]
> @rpatterson So, as far as I can see in this PR, the changes are "trivial": the main change is the introduction of the userLoggedIn as a signal that a user is authenticated, instead of having it based on the existence of the token. All the other changes are cosmetic, right?

Yeah, with the change of approach to do everything server-side in the PAS plugins and to preserve the JWT token, this PR is now essentially just all the cleanup, refactoring, and build improvements made along that way that I thought were still valuable.  Sorry for not spelling that out more clearly in the PR description.

**Comment by tiberiuichim** [comment]
Ok, so the related PRs in plone.volto and plone.restapi are not actually needed, correct? This could be merged into Volto without waiting for the others.

**Comment by rpatterson** [comment]
> Ok, so the related PRs in plone.volto and plone.restapi are not actually needed, correct? This could be merged into Volto without waiting for the others.

Oh, right.  Yeah, I misspoke, sorry.  I'll leave it up to you, but I've been testing locally against those PR branches so I've not confirmed that they're *not* needed, but AFAIK they shouldn't be needed.

**Comment by rpatterson** [comment]
Ok, the PRs this was built on top of have all been reviewed, approved and merged.  So AFAIK, this can be merged.  LMK if I'm missing anything.  CC: @tiberiuichim @sneridagh

File: CHANGELOG.md
Comment (tiberiuichim) [inline CHANGELOG.md:8]: Is this valid, in the context of Volto?

File: CHANGELOG.md
Comment (tiberiuichim) [inline CHANGELOG.md:1095]: I don't see the changelog entry matching the actual changes.

File: traefik.d/local.yml
Comment (tiberiuichim) [inline traefik.d/local.yml:1]: Should be moved to documentation.

**Comment by rpatterson** [comment]
For future reference, here's a `./Makefile` target I've been using to reproduce upgrade/migration issues with existing ZODBs:

    .PHONY: test-upgrades
    test-upgrades:  ## Test upgrading existing ZODBs
    # Start out before the PR changes
    	git checkout master
    # Delete and recreate the Zope instance ZODB
    	rm -r "./api/.installed.cfg" "./api/var/filestorage/" "./api/var/blobstorage/"
    	$(MAKE) build
    # Activate job control so back-grounded process get their own PGID so we can kill all
    # their descendant processes: https://stackoverflow.com/a/6552484/624787
    	set -m
    # Start the clean instance
    	$(MAKE) start &
    	set +x
    	until nc -vz "localhost" "3000"; do sleep 1; done
    	echo "In the browser, log into the Volto/React UI as the Zope root"
    	echo "'/acl_users' user 'admin' to ensure the browser has a JWT token."
    	read -p "Hit <ENTER> when done."
    	set -x
    	kill %1
    	while nc -vz "localhost" "3000"; do sleep 1; done
    # Force the backend Plone classic UI to be served on the same port so the browser will
    # send the same cookies, etc..
    	rm -v "./api/.installed.cfg"
    	$(MAKE) -C "./api/" INSTANCE_PORT=3000 start &
    	set +x
    	until nc -vz "localhost" "3000"; do sleep 1; done
    	echo "In the browser, visit the Zope root ZMI:"
    	echo "http://localhost:3000/manage_main"
    	read -p "Hit <ENTER> when done."
    	set -x
    	kill %1
    	while nc -vz "localhost" "3000"; do sleep 1; done
    # Checkout the branch that changes the plugins
    	git checkout fix-unify-auth
    # Start the backend again on the same port, but don't install the `plonesite` part to
    # avoid updating/upgrading the ZODB
    	cd "./api/"
    	"./bin/buildout" instance:http-address=3000 install instance
    	"./bin/instance" fg &
    	set +x
    	until nc -vz "localhost" "3000"; do sleep 1; done
    	echo "In the browser, visit the Zope root ZMI:"
    	echo "http://localhost:3000/manage_main"
    	read -p "Hit <ENTER> when done."
    	set -x
    	kill %1

File: CHANGELOG.md
Comment (rpatterson) [reply CHANGELOG.md]: Right you are, [pushed](https://github.com/plone/volto/commit/8d887af0e9c18b03effb86a99ade53f96bdec710).  Thanks for the catch.

File: CHANGELOG.md
Comment (rpatterson) [reply CHANGELOG.md]: [Ditto](https://github.com/plone/volto/pull/2959#discussion_r814652565)

**Comment by rpatterson** [comment]
> mag Inspect the deploy log: https://app.netlify.com/sites/volto/deploys/6218aeab225ab70008a1c0a6

The error message there is about a missing `./Makefile` target, `netlify`, but from the logs I can't see any changes of mine that could have removed such a target.  Can someone help me understand how my changes might have affected that or may I ignore that message?

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Minor grammar and typo fixes, and attempt to improve docstring for `arrayWIdsToObject`.

**Comment by stevepiercy** [comment]
> > mag Inspect the deploy log: https://app.netlify.com/sites/volto/deploys/6218aeab225ab70008a1c0a6
> 
> The error message there is about a missing `./Makefile` target, `netlify`, but from the logs I can't see any changes of mine that could have removed such a target. Can someone help me understand how my changes might have affected that or may I ignore that message?

It fails because the version of the `Makefile` in this PR does not include the `netlify` target that was added within the last couple of weeks to the `master` branch to preview the build of docs.

You could ignore it, but it might be a good idea to rebase your branch on `master`, then force push, or merge `master` onto your branch then push, in order to get any of the goodies that have come through in the last two months since opening this PR.

**Comment by rpatterson** [comment]
> It fails because the version of the `Makefile` in this PR does not include the `netlify` target that was added within the last couple of weeks to the `master` branch to preview the build of docs.

Doh!  I should have checked for new commits on master before asking for help there, sorry!  Thanks for the helping hand.  [I've merged](https://github.com/plone/volto/commit/ff43ee3dab0213c99d17a0b10e47eb7dba8a7ff3) and pushed.

File: src/helpers/Utils/Utils.js
Comment (rpatterson) [reply src/helpers/Utils/Utils.js]: Yeah, I struggled with this when I wrote that.  [I've taken a stab at improving using your wording as inspiration](https://github.com/plone/volto/commit/a5ccfd779329ff9dc0c5673314ada80eccaca72b).  Pushed.

Also, for some reason I can't reply to your comment below so I'll reply here:

> In arrayWIdsToObject, what does the W stand for?

"W" == "With"

File: src/helpers/Utils/Utils.js
Comment (rpatterson) [reply src/helpers/Utils/Utils.js]: Nice catch, thanks!  Pushed [the fix](https://github.com/plone/volto/commit/a5ccfd779329ff9dc0c5673314ada80eccaca72b).

File: docs/source/recipes/folder-structure.md
Comment (rpatterson) [reply docs/source/recipes/folder-structure.md]: > Avoid acronyms.

Why is that?  I think maximizing brevity without sacrificing clarity or completeness is valuable in documentation as it makes it more approachable, easier to scan, and generally more likely to be actually read by users/developers.  What is your reason for your opinion.

I tend to think feedback is best given with reasons.  Even if the 2 people involved both understand the why, it helps junior contributors who might read it.

Regardless, I don't want this to hold up this PR any further than it already has been, so I've pushed [the requested change](https://github.com/plone/volto/commit/a5ccfd779329ff9dc0c5673314ada80eccaca72b).

Comment (stevepiercy) [reply docs/source/recipes/folder-structure.md]: I should have clarified. Avoid _non-technical_ acronyms. It's fine to use "HTML", "CSS", and so on, but "AKA" is not technical. Less common technical acronyms should also be defined when first introduced in narrative text ("vehicle miles travelled (VMT) is a measure of the mileage covered along a stretch of roadway by motorized vehicles"), and added to the Glossary and use `{term}` to reference the Glossary entry.

As an aside, "AKA"'s proper usage is either "a.k.a." or "aka", depending on the context. Legal and criminal contexts generally use "aka".

In general English is not the first language of many developers, and that is even more true in Plone and Volto. It is always better to spell things out, and avoid abbreviations and English idioms. Google Translate does a lot better without them.

**Review by stevepiercy — COMMENTED** [review]
Suggestions for improving the function name.

**Comment by stevepiercy** [comment]
I pushed commit https://github.com/plone/volto/pull/2959/commits/ffc731ed4480eefab37fada07e907489d788d903 for review.

I also commented on the function name: https://github.com/plone/volto/pull/2959#discussion_r817466767

File: src/helpers/Utils/Utils.js
Comment (rpatterson) [reply src/helpers/Utils/Utils.js:205]: None of those seem to be clear improvements to me.  If any of them get an up-vote from anyone else I'll make the change just to finish this discussion.  Until then I'm resolving this.

File: src/helpers/Utils/Utils.js
Comment (rpatterson) [inline src/helpers/Utils/Utils.js:190]: I think the first line should be a standalone sentence, much like Python docstrings, git commit messages, etc..  But I don't care more about that than I care about finishing this discussion.

File: docs/source/recipes/folder-structure.md
Comment (rpatterson) [reply docs/source/recipes/folder-structure.md]: > In general English is not the first language of many developers, and that is even more true in Plone and Volto. It is always better to spell things out, and avoid abbreviations and English idioms. Google Translate does a lot better without them.

Ah, those are reasons that sell for me.

**Review by avoinea — APPROVED** [review]
LGTM. 

Tested on localhost and demo website with LDAP:
- Plone: 6.0.0a3
- PlonePAS: 7.0.0a3
- plone.restapi "fix-unify-auth-logout":https://github.com/plone/plone.restapi/pull/1303
- plone.restapi "fix-zope-root-auth":https://github.com/plone/plone.restapi/pull/1304

**Review by rpatterson — COMMENTED** [review]
I can't find any discussion of the why for this change.  Am I missing that somewhere?

**Pull request closed without merging**
